src/index.ts

```diff
@@ -0,0 +1,193 @@
+
+    // record current state, which is used to initialize new
+    // subscribers of this topic with the last published state
+    topicRecord.currentState = newState
+
+    // set a flag to indicate that this topic has been published to
+    topicRecord.hasBeenPublished = true
+
+  } else {
+
+    throw new Error(`Cannot publish to non-existent topic "${topic}"`)
+
+  }
+
+}
+
+/**
+ * Sets up the unsubscribe function that is returned for a specific topi subscriber.
+ */
+const unsubscribe: InternalUnsubscribe = (topic: string, subscriptionId?: string) => () => {
+
+  const topicRecord = topics[topic]
+
+  if (topicRecord && subscriptionId) {
+
+    delete topicRecord.subscriptionMap[subscriptionId]
+
+  }
+
+}
+
+/**
+ * Hook that enables pub-sub functionality across ReactJS function components.
+ */
+function usePubSub<T>(topic: string, defaultState: T): SubscriptionTuple<T> {
+
+  if (arguments.length !== 2) {
+
+    throw new Error(
+      `Invalid hook usage; hook must be initialized with two arguments. The first
+      argument is the topic and the second argument is the default state value for the topic.`
+    )
+
+  }
+
+  // internal hook that actually does the dispatching to this new subscriber
+  const [state, setState] = useState<T>()
+
+  // internal hook that stores the unique ID for this topic subscriber
+  const [subscriptionId, setSubscriptionId] = useState<string>()
+
+  /**
+   * Internal hook for this subscription that is used to cache the resulting tuple.
+   */
+  const internalUsePubSubState = () => {
+
+    let currentState = getCurrentState(topic)
+
+    currentState = typeof currentState !== 'undefined'
+      ? currentState = currentState as SetStateAction<T>
+      : currentState = defaultState
+
+    if (typeof state === 'undefined' && typeof currentState !== 'undefined') {
+      setState(currentState)
+    }
+
+    return [state, setState] as InternalTuple<T>
+
+  }
+
+  let internalTuple: InternalTuple<T>
+
+  if (!subscriptionId) {
+
+    internalTuple = internalUsePubSubState()
+
+  }
+
+  useEffect(() => {
+
+    //
+    // create the new subscription when subscribing component mounts
+    //
+
+    const newSubscriptionId = getUUID()
+
+    setSubscriptionId(newSubscriptionId)
+
+    if (!topics[topic]) {
+
+      //
+      // initialize a new topic record for topic
+      //
+      topics[topic] = {
+        currentState: state,
+        hasBeenPublished: false,
+        subscriptionMap: {},
+      }
+
+    }
+
+    // store tuple for new subscription
+    topics[topic].subscriptionMap[newSubscriptionId] = internalTuple
+
+    return () => {
+
+      //
+      // automatically unsubscribe from topic when
+      // component unmounts (to prevent memory leaks)
+      //
+      unsubscribe(topic, newSubscriptionId)()
+
+    }
+
+  }, [])
+
+  // return the subscription tuple
+  return [state || defaultState, publish<T>(topic), unsubscribe(topic, subscriptionId)]
+
+}
+
+export default usePubSub
```

Comment:
Can we start dumping default exports in favor of named exports, letting folks `import { bar as foo }` if they want to rename? `goto declaration` and some linting auto fix features just work better with named exports (especially in the lerna mono-repo env).

Reply:
Sounds reasonable to me. I went ahead and snuck this one in on my branch and did a quick publish so that the interface is correct (in case anyone starts downloading).

Will wait a while to push to PR to allow more time for review comments. Great ask though.